schemas/telemetry/untrusted-modules/untrusted-modules.4.parquetmr.txt

```diff
@@ -0,0 +1,313 @@
+      optional int64 cores;
+      optional int64 count;
+      optional int64 family (UINT_64);
+      optional int64 l2cacheKB (UINT_64);
+      optional int64 l3cacheKB (UINT_64);
+      optional int64 model (UINT_64);
+      optional int64 speedMHz (UINT_64);
+      optional int64 stepping (UINT_64);
+      optional binary vendor (UTF8);
+      optional group extensions (LIST) {
+        repeated group list {
+          required binary element (UTF8);
+        }
+      }
+    }
+    required group build {
+      required binary applicationId (UTF8);
+      required binary applicationName (UTF8);
+      required binary architecture (UTF8);
+      optional binary architecturesInBinary (UTF8);
+      required binary buildId (UTF8);
+      optional binary hotfixVersion (UTF8);
+      required binary version (UTF8);
+      required binary vendor (UTF8);
+      optional binary displayVersion (UTF8);
+      required binary platformVersion (UTF8);
+      optional boolean updaterAvailable;
+      required binary xpcomAbi (UTF8);
+    }
+    required group partner {
+      optional binary distributionId (UTF8);
+      optional binary distributionVersion (UTF8);
+      optional binary partnerId (UTF8);
+      optional binary distributor (UTF8);
+      optional binary distributorChannel (UTF8);
+      optional group partnerNames (LIST) {
+        repeated group list {
+          optional binary element (UTF8);
+        }
+      }
+    }
+    required group addons {
+      optional binary persona (UTF8);
+      required group theme {
+        required binary id (UTF8);
+        optional boolean blocklisted;
+        optional binary description (UTF8);
+        optional binary name (UTF8);
+        optional boolean userDisabled;
+        optional boolean appDisabled;
+        optional binary version (UTF8);
+        optional int64 scope;
+        optional boolean hasBinaryComponents;
+        optional int64 installDay (UINT_64);
+        optional int64 updateDay (UINT_64);
+      }
+      required group activeAddons (MAP) {
+        repeated group key_value {
+          required binary key (UTF8);
+          required group value {
+            required boolean blocklisted;
+            required binary name (UTF8);
+            optional binary description (UTF8);
+            required boolean appDisabled;
+            required int64 scope;
+            required binary type (UTF8);
+            required boolean hasBinaryComponents;
+            required int64 installDay (UINT_64);
+            required int64 updateDay (UINT_64);
+            optional int64 signedState;
+            required boolean isSystem;
+          }
+        }
+      }
+      required group activeGMPlugins (MAP) {
+        repeated group key_value {
+          required binary key (UTF8);
+          required group value {
+            required binary version (UTF8);
+            required boolean userDisabled;
+          }
+        }
+      }
+      required group activePlugins (LIST) {
```

Comment:
Seems this can be either a list or an object [0], but the docs don't indicate when it's not a list. Any ideas? If it ever is returned an object the ping will be discarded.

[0] https://github.com/mozilla-services/mozilla-pipeline-schemas/blob/dev/templates/include/telemetry/environment.1.schema.json#L93

Reply:
Hmm, looking at where we generate this, it is always returned as an array.

https://searchfox.org/mozilla-central/rev/4e094f66ced333d69b24cd49273789e3a1173dfc/toolkit/components/telemetry/app/TelemetryEnvironment.jsm#764

But as above I would prefer to be perfectly correct here, especially as fields in the environment block are really a nice-to-have for us. If there's a risk that we're filtering out good data because of this, I'd prefer to go safe than sorry.

What do you think?